### DISCUSSION TITLE
Add policy based on jwt

### DISCUSSION THREAD

**Discussion by Tarang** [supabase/supabase#824] [Questions]
Hi,

I would like to mint my own JWT and not use Supabases auth, or use it as minimally as possible as it misses a lot of the functionality I need. Using the minted JWT how is it possible to use some of the jwt fields to create a policy?

e.g the jwt payload could be `{ "id": [12,13,14] }` and so the rows with id 12, 13 or 14 should be allowed only.

Thanks

**Comment by kiwicopple** [comment]
You can write functions which access `request.jwt.claim`. For example, this is the function we use for `auth.uid()` to get the ID from the JWT:

```sql
create or replace function auth.uid() returns uuid as $$
  select nullif(current_setting('request.jwt.claim.sub', true), '')::uuid;
$$ language sql stable;
```

Hope that helps @Tarang ! let us know if you need anything else

**Reply by Tarang** [reply]
Thats fantastic! Do you have to have a user in the users table for this? Also are there more docs/examples on how to write policies anywhere?

Thanks @kiwicopple

**Reply by kiwicopple** [reply]
No you don't need any users - you can create these functions anywhere.

> Also are there more docs/examples

We don't have too much (yet) but we're working on a lot more. @awalias did a fantastic series on Auth here https://supabase.io/docs/learn/auth-deep-dive/auth-row-level-security.

Also note that we're just using Postgres, so you've got a wealth of docs online, and also we use PostgREST (which is amazing), and has some great docs on Auth here: https://postgrest.org/en/v7.0.0/auth.html#client-auth

**Reply by Tarang** [reply]
thanks @kiwicopple I can see it working. I was wondering what is the purpose of the api key if there is an authorisation header? Is there a way to remove it if its redundant or to mint my own tokens there?

**Reply by Tarang** [reply]
Also @kiwicopple  how does the json work.

Say I have my token and its payload is `{ "id": [12,13,14] }`, how can I get it to parse the json array with your example:
```select nullif(current_setting('request.jwt.claim.id', true), '')::int[];```

I'm a bit unsure of how json is interpreted here, is `current_setting('request.jwt.claim.id', true)` of type `json`?
